gwctl/pkg/policymanager/manager.go

```diff
@@ -0,0 +1,345 @@
+/*
+Copyright 2023 The Kubernetes Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package policymanager
+
+import (
+	"context"
+	"encoding/json"
+	"fmt"
+
+	apiextensionsv1 "k8s.io/apiextensions-apiserver/pkg/apis/apiextensions/v1"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/apis/meta/v1/unstructured"
+	"k8s.io/apimachinery/pkg/runtime"
+	"k8s.io/apimachinery/pkg/runtime/schema"
+	"k8s.io/client-go/dynamic"
+	gatewayv1alpha2 "sigs.k8s.io/gateway-api/apis/v1alpha2"
+)
+
+const (
+	gatewayPolicyLabelKey = "gateway.networking.k8s.io/policy"
+)
+
+type PolicyManager struct {
+	dc dynamic.Interface
+
+	// policyCRDs maps a CRD name to the CRD object.
+	policyCRDs map[PolicyCrdID]PolicyCRD
+	// policies maps a policy name to the policy object.
+	policies map[string]Policy
+}
+
+func New(dc dynamic.Interface) *PolicyManager {
+	return &PolicyManager{
+		dc:         dc,
+		policyCRDs: make(map[PolicyCrdID]PolicyCRD),
+		policies:   make(map[string]Policy),
+	}
+}
+
+// Init will construct a local cache of all Policy CRDs and Policy Resources.
+func (p *PolicyManager) Init(ctx context.Context) error {
+	allCRDs, err := fetchCRDs(ctx, p.dc)
+	if err != nil {
+		return err
+	}
+	for _, crd := range allCRDs {
+		policyCRD := PolicyCRD{crd}
+		// Check if the CRD is a Gateway Policy CRD
+		if policyCRD.IsValid() {
+			p.policyCRDs[policyCRD.ID()] = policyCRD
+		}
+	}
+
+	allPolicies, err := fetchPolicies(ctx, p.dc, p.policyCRDs)
+	if err != nil {
+		return err
+	}
+	for _, unstrucutredPolicy := range allPolicies {
+		policy, err := PolicyFromUnstrucutred(unstrucutredPolicy, p.policyCRDs)
+		if err != nil {
+			return err
+		}
+		p.policies[unstrucutredPolicy.GetNamespace()+"/"+unstrucutredPolicy.GetName()] = policy
```

Comment:
This may benefit from a separate function like `p.AddPolicy(unstructuredPolicy)`

Reply:
Nice idea. Thanks changed.